hack/lib/util/trap.sh

```diff
@@ -0,0 +1,269 @@
+#  - export REGISTERED_TRAPS_TO_SIGNALS
+#  - export SIGNALS_TO_REGISTERED_TRAPS
+function os::util::trap::internal::init() {
+    # REGISTERED_TRAPS_TO_SIGNALS is an associative array of traps to the signals for which they are registered
+    declare -A REGISTERED_TRAPS_TO_SIGNALS
+    export REGISTERED_TRAPS_TO_SIGNALS
+    # SIGNALS_TO_REGISTERED_TRAPS is an associative array of signals to the traps registered for them
+    declare -A SIGNALS_TO_REGISTERED_TRAPS
+    export SIGNALS_TO_REGISTERED_TRAPS
+
+    # When sourced we need this script to pre-emptively register all possible traps.
+    os::trap::internal::register_traps
+}
+
+# os::util::trap::add appends a command to the list of traps for any signals specified. 
+# This method should be used as the default trap addition method.
+#
+# Globals:
+#  None
+# Arguments:
+#  - 1: command to trap
+#  - 2+: signals to add trap to
+# Returns:
+#  None
+function os::util::trap::add() {
+    local trap_command=$1; shift
+    local signals=$@
+
+    for signal in ${signals}; do
+        os::util::trap::append_to_signal "${trap_command}" "${signal}"
+    done
+}
+
+# os::util::trap::remove removes the named trap from any signals that it's registered under
+# 
+# Globals:
+#  - SIGNALS_TO_REGISTERED_TRAPS
+#  - REGISTERED_TRAPS_TO_SIGNALS
+# Arguments
+#  - 1: the trap handler to remove
+# Returns
+#  - update SIGNALS_TO_REGISTERED_TRAPS
+#  - update REGISTERED_TRAPS_TO_SIGNALS 
+function os::util::trap::remove() {
+    local trap_command=$1
+
+    local signals
+    signals="${REGISTERED_TRAPS_TO_SIGNALS[${trap_command}]}"
+
+    for signal in ${signals}; do
+        local pruned_list
+        for registered_trap in ${SIGNALS_TO_REGISTERED_TRAPS["${signal}"]-}; do
+            if [[ ! "${trap_command}" == "${registered_trap}" ]]; then
+                # we keep any registered traps for this signal that aren't the trap we remove
+                pruned_list="${pruned_list} ${registered_trap}"
+            fi
+        done
+
+        SIGNALS_TO_REGISTERED_TRAPS["${signal}"]="${pruned_list}"
+    done
+
+    unset REGISTERED_TRAPS_TO_SIGNALS["${trap_command}"]
+}
+
+# os::util::trap::append_to_signal appends the trap command to the list of trap handlers for the signal.
+#
+# Globals:
+#  - SIGNALS_TO_REGISTERED_TRAPS
+#  - REGISTERED_TRAPS_TO_SIGNALS
+# Arguments:
+#  - 1: command to trap
+#  - 2: signal to trap
+# Returns:
+#  - update SIGNALS_TO_REGISTERED_TRAPS
+#  - update REGISTERED_TRAPS_TO_SIGNALS
+function os::util::trap::append_to_signal() {
+    local trap_command=$1
+    local signal=$2
+
+    signal="$(os::util::trap::internal::prefix_signal_if_necessary "${signal}")"
+
+    os::util::trap::internal::validate_trap "${trap_command}"
+    os::util::trap::internal::validate_signal "${signal}"
+
+    # update the associative arrays with the new registration if we haven't already done the registration
+    if ! echo "${SIGNALS_TO_REGISTERED_TRAPS[${signal}]-}" | grep -q "${trap_command}"; then
```

Comment:
@eparis if you're reviewing, please take a look at this. I can't explain this behavior ...
The key is evaluated, `${signal}` --> `ERR` or something.
Then, the *value* is evaluated *again*, causing an error saying that `ERR` is an unset variable due to `set -o nounset`.

Log here:
```
++++ os::util::trap::add func ERR
++++ local trap_command=func
++++ shift
++++ local signals=ERR
++++ for signal in '\''${signals}'\''
++++ os::util::trap::append_to_signal func ERR
++++ local trap_command=func
++++ local signal=ERR
+++++ os::util::trap::internal::prefix_signal_if_necessary ERR
+++++ local raw_signal=ERR
+++++ local signals
++++++ os::trap::internal::list_all_possible_signals
++++++ local signals
++++++ signals=($(kill -l | grep -Po "(?<=[0-9]\) )[^\t ]+"))
+++++++ kill -l
+++++++ grep -Po '\''(?<=[0-9]\) )[^\t ]+'\''
++++++ signals+=("EXIT" "ERR")
++++++ echo SIGHUP SIGINT SIGQUIT SIGILL SIGTRAP SIGABRT SIGBUS SIGFPE SIGKILL SIGUSR1 SIGSEGV SIGUSR2 SIGPIPE SIGALRM SIGTERM SIGSTKFLT SIGCHLD SIGCONT SIGSTOP SIGTSTP SIGTTIN SIGTTOU SIGURG SIGXCPU SIGXFSZ SIGVTALRM SIGPROF SIGWINCH SIGIO SIGPWR SIGSYS SIGRTMIN SIGRTMIN+1 SIGRTMIN+2 SIGRTMIN+3 SIGRTMIN+4 SIGRTMIN+5 SIGRTMIN+6 SIGRTMIN+7 SIGRTMIN+8 SIGRTMIN+9 SIGRTMIN+10 SIGRTMIN+11 SIGRTMIN+12 SIGRTMIN+13 SIGRTMIN+14 SIGRTMIN+15 SIGRTMAX-14 SIGRTMAX-13 SIGRTMAX-12 SIGRTMAX-11 SIGRTMAX-10 SIGRTMAX-9 SIGRTMAX-8 SIGRTMAX-7 SIGRTMAX-6 SIGRTMAX-5 SIGRTMAX-4 SIGRTMAX-3 SIGRTMAX-2 SIGRTMAX-1 SIGRTMAX EXIT ERR
+++++ signals='\''SIGHUP SIGINT SIGQUIT SIGILL SIGTRAP SIGABRT SIGBUS SIGFPE SIGKILL SIGUSR1 SIGSEGV SIGUSR2 SIGPIPE SIGALRM SIGTERM SIGSTKFLT SIGCHLD SIGCONT SIGSTOP SIGTSTP SIGTTIN SIGTTOU SIGURG SIGXCPU SIGXFSZ SIGVTALRM SIGPROF SIGWINCH SIGIO SIGPWR SIGSYS SIGRTMIN SIGRTMIN+1 SIGRTMIN+2 SIGRTMIN+3 SIGRTMIN+4 SIGRTMIN+5 SIGRTMIN+6 SIGRTMIN+7 SIGRTMIN+8 SIGRTMIN+9 SIGRTMIN+10 SIGRTMIN+11 SIGRTMIN+12 SIGRTMIN+13 SIGRTMIN+14 SIGRTMIN+15 SIGRTMAX-14 SIGRTMAX-13 SIGRTMAX-12 SIGRTMAX-11 SIGRTMAX-10 SIGRTMAX-9 SIGRTMAX-8 SIGRTMAX-7 SIGRTMAX-6 SIGRTMAX-5 SIGRTMAX-4 SIGRTMAX-3 SIGRTMAX-2 SIGRTMAX-1 SIGRTMAX EXIT ERR'\''
+++++ echo '\''SIGHUP SIGINT SIGQUIT SIGILL SIGTRAP SIGABRT SIGBUS SIGFPE SIGKILL SIGUSR1 SIGSEGV SIGUSR2 SIGPIPE SIGALRM SIGTERM SIGSTKFLT SIGCHLD SIGCONT SIGSTOP SIGTSTP SIGTTIN SIGTTOU SIGURG SIGXCPU SIGXFSZ SIGVTALRM SIGPROF SIGWINCH SIGIO SIGPWR SIGSYS SIGRTMIN SIGRTMIN+1 SIGRTMIN+2 SIGRTMIN+3 SIGRTMIN+4 SIGRTMIN+5 SIGRTMIN+6 SIGRTMIN+7 SIGRTMIN+8 SIGRTMIN+9 SIGRTMIN+10 SIGRTMIN+11 SIGRTMIN+12 SIGRTMIN+13 SIGRTMIN+14 SIGRTMIN+15 SIGRTMAX-14 SIGRTMAX-13 SIGRTMAX-12 SIGRTMAX-11 SIGRTMAX-10 SIGRTMAX-9 SIGRTMAX-8 SIGRTMAX-7 SIGRTMAX-6 SIGRTMAX-5 SIGRTMAX-4 SIGRTMAX-3 SIGRTMAX-2 SIGRTMAX-1 SIGRTMAX EXIT ERR'\''
+++++ grep -q SIGERR
+++++ echo ERR
++++ signal=ERR
++++ os::util::trap::internal::validate_trap func
++++ local trap_command=func
++++ echo func
++++ grep -q '\''^[^[:space:]\'\''\'\'''\''"]+$'\''
++++ os::util::trap::internal::validate_signal ERR
++++ local signal=ERR
++++ local signals
+++++ os::trap::internal::list_all_possible_signals
+++++ local signals
+++++ signals=($(kill -l | grep -Po "(?<=[0-9]\) )[^\t ]+"))
++++++ kill -l
++++++ grep -Po '\''(?<=[0-9]\) )[^\t ]+'\''
+++++ signals+=("EXIT" "ERR")
+++++ echo SIGHUP SIGINT SIGQUIT SIGILL SIGTRAP SIGABRT SIGBUS SIGFPE SIGKILL SIGUSR1 SIGSEGV SIGUSR2 SIGPIPE SIGALRM SIGTERM SIGSTKFLT SIGCHLD SIGCONT SIGSTOP SIGTSTP SIGTTIN SIGTTOU SIGURG SIGXCPU SIGXFSZ SIGVTALRM SIGPROF SIGWINCH SIGIO SIGPWR SIGSYS SIGRTMIN SIGRTMIN+1 SIGRTMIN+2 SIGRTMIN+3 SIGRTMIN+4 SIGRTMIN+5 SIGRTMIN+6 SIGRTMIN+7 SIGRTMIN+8 SIGRTMIN+9 SIGRTMIN+10 SIGRTMIN+11 SIGRTMIN+12 SIGRTMIN+13 SIGRTMIN+14 SIGRTMIN+15 SIGRTMAX-14 SIGRTMAX-13 SIGRTMAX-12 SIGRTMAX-11 SIGRTMAX-10 SIGRTMAX-9 SIGRTMAX-8 SIGRTMAX-7 SIGRTMAX-6 SIGRTMAX-5 SIGRTMAX-4 SIGRTMAX-3 SIGRTMAX-2 SIGRTMAX-1 SIGRTMAX EXIT ERR
++++ signals='\''SIGHUP SIGINT SIGQUIT SIGILL SIGTRAP SIGABRT SIGBUS SIGFPE SIGKILL SIGUSR1 SIGSEGV SIGUSR2 SIGPIPE SIGALRM SIGTERM SIGSTKFLT SIGCHLD SIGCONT SIGSTOP SIGTSTP SIGTTIN SIGTTOU SIGURG SIGXCPU SIGXFSZ SIGVTALRM SIGPROF SIGWINCH SIGIO SIGPWR SIGSYS SIGRTMIN SIGRTMIN+1 SIGRTMIN+2 SIGRTMIN+3 SIGRTMIN+4 SIGRTMIN+5 SIGRTMIN+6 SIGRTMIN+7 SIGRTMIN+8 SIGRTMIN+9 SIGRTMIN+10 SIGRTMIN+11 SIGRTMIN+12 SIGRTMIN+13 SIGRTMIN+14 SIGRTMIN+15 SIGRTMAX-14 SIGRTMAX-13 SIGRTMAX-12 SIGRTMAX-11 SIGRTMAX-10 SIGRTMAX-9 SIGRTMAX-8 SIGRTMAX-7 SIGRTMAX-6 SIGRTMAX-5 SIGRTMAX-4 SIGRTMAX-3 SIGRTMAX-2 SIGRTMAX-1 SIGRTMAX EXIT ERR'\''
++++ echo '\''SIGHUP SIGINT SIGQUIT SIGILL SIGTRAP SIGABRT SIGBUS SIGFPE SIGKILL SIGUSR1 SIGSEGV SIGUSR2 SIGPIPE SIGALRM SIGTERM SIGSTKFLT SIGCHLD SIGCONT SIGSTOP SIGTSTP SIGTTIN SIGTTOU SIGURG SIGXCPU SIGXFSZ SIGVTALRM SIGPROF SIGWINCH SIGIO SIGPWR SIGSYS SIGRTMIN SIGRTMIN+1 SIGRTMIN+2 SIGRTMIN+3 SIGRTMIN+4 SIGRTMIN+5 SIGRTMIN+6 SIGRTMIN+7 SIGRTMIN+8 SIGRTMIN+9 SIGRTMIN+10 SIGRTMIN+11 SIGRTMIN+12 SIGRTMIN+13 SIGRTMIN+14 SIGRTMIN+15 SIGRTMAX-14 SIGRTMAX-13 SIGRTMAX-12 SIGRTMAX-11 SIGRTMAX-10 SIGRTMAX-9 SIGRTMAX-8 SIGRTMAX-7 SIGRTMAX-6 SIGRTMAX-5 SIGRTMAX-4 SIGRTMAX-3 SIGRTMAX-2 SIGRTMAX-1 SIGRTMAX EXIT ERR'\''
++++ grep -q '\''^ERR$'\''
hack/test/../../hack/lib/util/trap.sh: line 107: ERR: unbound variable
++++ grep -q func
hack/test/../../hack/lib/util/trap.sh: line 108: ERR: unbound variable'
```